### PR TITLE
Fix information schema queries when cluster state has outdated views

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Relates to https://github.com/crate/crate/issues/14377

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java [hunk 1]
```diff
@@ -79,13 +78,10 @@ public ViewInfo create(RelationName ident, ClusterState state) {
...
-                // Return ViewInfo with no columns in case the statement could not be analyzed,
-                // because the underlying table of the view could not be found.
-                columns = Collections.emptyList();
-            } else {
-                throw e;
-            }
+            // Statement could not be analyzed, because the referenced table either not found
+            // or has been updated and view definition became incompatible with the new schema (https://github.com/crate/crate/issues/14377).
+            columns = Collections.emptyList();
+
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:84]: Note, that information schema would be working and "corrupted" view won't be displayed. 

Drawback of such solution (including ignoring `RelationUnknown` exception on master) is having unusable views in the cluster state.  `select * from outdated_view` would raise the same casting error. 

Another ways to fix it:

1. Keep views in sync with tables. I'm not sure it's possible and practical in all scenarios. In the reported issue user wanted to subtract an interval from a timestamp and expected to get a timestamp. We can force view to return array with a single value having the "expected" result but such view can, in turn, break other user queries, dependant on that view.

2.  Ignore all exceptions but also drop the corrupted view as it's anyway unusable. It can also break user's dependant queries (also changing cluster state when accessing information schema is not really safe).

3. Ignore exceptions but maybe log a warning on information schemas queries (view X is outdated and cannot be accessed) and show a better error message on `select * from outdated_view`?

Comment (proddata) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:84]: > Drawback of such solution (including ignoring RelationUnknown exception on master) is having unusable views in the cluster state.

If it can be fixed e.g. by `CREATE OR REPLACE VIEW ....` I think that is fine.

ad 1) I think that is also bad in table replacement scenarios that we rely on, as reindexing is sometimes needed
ad 2) see 1

ad 3) Instead of a log message (which is limited to host admins) I would rather have an error message in `information_schema.views`

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:84]: > have an error message in information_schema.views

had a concern that PG doesn't have anything like that but we already have our custom `owner` column and in general custom column should be fine in terms of compatibility. Sounds fine to  me, @matriv, what do you think?

Comment (matriv) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:84]: I think that this PR's solution is the way to go. The view is erroneous when you try to query it and then you can still see what is the query stored from the view through `information_schema.views` and `pg_views` and fix it.

- I wouldn't add extra cols to `information_schema.views` which is SQL standard.
- Adding a log message when querying those catalog info tables, I think it's not helpful for the user.

I have another suggestion, how about adding a comment like:
``-- Corrupeted - needs fix --`` on top of the query of the view, returned in those tables?

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:84]: > I have another suggestion, how about adding a comment like:
-- Corrupted - needs fix -- on top of the query of the view, returned in those tables?

Will do, thanks! Should we mention a better error message on broken queries in the entry?

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java]: Thanks, added a hint and updated a change entry.

File: server/src/test/java/io/crate/integrationtests/ViewsITest.java [hunk 2]
```diff
@@ -55,6 +58,36 @@ public void dropViews() {
...
+        execute("CREATE TABLE t01 (ts ARRAY(TIMESTAMP))");
+
+        // Verify that information schema is accessible
+        // and view definition has "corrupted, needs fix" hint as a comment.
+
+        execute("SELECT view_definition FROM information_schema.views where table_name = 'v01'");
+        assertThat(response).hasRows(
+            "/* Corrupted view, needs fix */ SELECT CAST(\"ts\" AS timestamp) - CAST('1 hour' AS interval)",
+            "FROM \"t01\""
+        );
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/ViewsITest.java:76]: I probably misunderstand syntax of hasRows params but to me it looks like we expect 2 rows. But in runtime it's a single row with \n before FROM.

Found an [existing example](https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/ViewsITest.java#L70-L73) where we do such assertion, also use multiple rows, probably it's fine.

File: server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java [hunk 3]
```diff
@@ -79,14 +80,12 @@ public ViewInfo create(RelationName ident, ClusterState state) {
...
-            } else {
-                throw e;
-            }
+            // Statement could not be analyzed, because the referenced table either not found
+            // or has been updated and view definition became incompatible with the new schema (https://github.com/crate/crate/issues/14377).
+            columns = Collections.emptyList();
+            analyzeError = true;
         }
-        return new ViewInfo(ident, view.stmt(), columns, view.owner());
+        String viewDefinition = analyzeError ? String.format(Locale.ENGLISH, "/* Corrupted view, needs fix */ %s", view.stmt()) : view.stmt();
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:88]: Didn't use -- as it makes comments the whole statement, I tried locally enclosing with `-- comment -- SQL` and and didn't work.

File: docs/appendices/release-notes/5.3.5.rst [hunk 4]
```diff
@@ -80,3 +80,8 @@ Fixes
   leave behind empty partitions if ``NULL`` or ``CHECK`` constraint on
   partitioned by column failed.
 
+- Fixed an issue which caused errors on querying ``information_schema`` tables
```
Comment (matriv) [inline docs/appendices/release-notes/5.3.5.rst:83]: I would rephrase it a bit to make it more generic:
```
Fixed an issue which caused errors on querying ``information_schema`` tables
when the query of a ``VIEW`` is erroneous due to changes made to the underlying tables/views. Also, added a comment to view definition in ``pg_catalog.pg_views`` and ``information_schema.views`` tables to denote that a ``VIEW``'s query is erroneous.
```

File: server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java (hunk 3 above)
Comment (matriv) [inline server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:88]: How about adding a line break to make the comment even more visible.
```suggestion
        String viewDefinition = analyzeError ? String.format(Locale.ENGLISH, "/* Corrupted view, needs fix */\n%s", view.stmt()) : view.stmt();
```

Comment (matriv) [reply server/src/main/java/io/crate/metadata/view/ViewInfoFactory.java:88]: Because you'd need a `\n` in between the comment and the query.

**Review by matriv — COMMENTED** [review]
Thank you! Left a couple of comments, also: could you please check if there is any possibility to have a unit test instead, by using the `ViewInfoFactory` (like for example in `DocIndexMetadataTest`)?

**Comment by BaurzhanSakhariev** [comment]
> if there is any possibility to have a unit test instead

I added a unit test and removed Itest. I wanted to ensure that information schema is accessible but probably unit test is enough since we know that for that specific issue throwing an exception was the root cause.

File: server/src/test/java/io/crate/metadata/view/ViewInfoFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        ViewsMetadata viewsMetadata = mock(ViewsMetadata.class);
+        String statement = "SELECT * FROM users";
+        RelationName ident = new RelationName(null, "test");
+        when(viewsMetadata.getView(ident)).thenReturn(new ViewMetadata(statement, null));
+
+        ClusterState state  = ClusterState.builder(ClusterState.EMPTY_STATE)
+            .metadata(Metadata.builder().putCustom(ViewsMetadata.TYPE, viewsMetadata)).build();
+
+        // Selecting from view itself throws an error.
```
Comment (matriv) [inline server/src/test/java/io/crate/metadata/view/ViewInfoFactoryTest.java:52]: Please adjust: 
```suggestion
        // `definition` col includes a hint about an error in the view's query
```

**Review by matriv — APPROVED** [review]
Thx!

**Review by matriv — APPROVED** [review]

**Review by matriv — APPROVED** [review]
